### ISSUE TITLE
setting simulator to run on Spinnaker

### ISSUE THREAD

**Issue by piewchee** [NeuromorphicProcessorProject/snn_toolbox#23]
Hi Bodo,
You have a great toolbox!
I understand that the toolbox can use Spinnaker as simulator after converting to SNN.
May I know if I wish to run the SNN on Spinnaker, what should I change in the config file ?
simulator = spiNNaker?

I tried changing the simulator = nest.
I am unable to run due to ImportError: no module named nest.

May I know what else should I change in order for it to run?

To think that the pyNN_target_sim.py is quite similar to Spinnaker. Not sure what to modify
in order to be compatible with Spinnaker.

Best Regards
Del

**Comment by rbodo** [comment]
Hi Del,

Thanks for your interest in our toolbox! Sadly, the toolbox is not yet interfaced directly with SpiNNaker. 

First you need to install one of the pyNN simulators, preferably nest (faster than brian in my experience).  

Second, set `simulator = nest`. Your previous error message should not appear now.

As part of the output, the toolbox then provides you with an SNN model in pyNN format which you can use for SpiNNaker. Specifically, after building the SNN, the toolbox saves a text file for each layer containing the connection source- and target-indices and the corresponding weight and delay.

Hope this helps.

Best,

Bodo

**Comment by piewchee** [comment]
Hi Bodo,

Thanks for the reply.
When I use nest as simulator for Lenet5/Keras.
the moving accuray for ANN was 100%, 100%
whereas SNN was 0%, 11.1%
the SNN total accuracy is 0% for 10 test examples.
May I know what is wrong?

When I run INI,
the SNN total accuracy is 80%.

Rgds
Del

**Comment by piewchee** [comment]
> 
> 
> There are a few things to try here. (To give you more specific feedback, I'd need to see the config file.)
> 
>     * Increase the simulation duration. Compared to INI simulator, pyNN simulators should be given about 3 times as long to settle on stable firing rates.
> 
>     * Make sure parameter normalization is enabled in the config file.
> 
>     * Before trying to fix the SNN accuracy with pyNN simulators, you should get good accuracy with INI simulator first. 80% shows that something's wrong in the pipeline (maybe already during parsing of the input model, or because parameter normalization is disabled). Make sure that the accuracy of the model is about 98% after each stage of the pipeline (input model, parsed & normalized model, converted model run with INI sim). Then switch to Nest simulator.
> 
>     * Check the output plots. The SNN spikerates should correspond closely to the ANN activations in each feature map. Make sure the spikerates are not too low overall, or on the contrary, saturated. If they are, you can play with the parameter normalization.

Hi Bodo,

I did not change the config file parameters in keras, which I git clone.
In the config.defaults file,  online normalisation and schedule are false.
the keras config. file batch_size = 1, increase this will help?

Rgds
Del

**Comment by rbodo** [comment]
OK. You don't have to worry about `online normalization` or `normalization schedule`. If you did not change anything, parameter normalization should be enabled by default. It's the setting 

```
[tools]
normalize = True
```

I don't think the batch size will make a difference. 

I will run the example later today to see if I can reproduce the behavior.

**Comment by rbodo** [comment]
Did you try with a longer simulation time (e.g. 100 time steps)? This should give you good accuracy on the LeNet example with Nest simulator. (Even though the ANN accuracy is still low - which is due to the model being trained with an old Keras version that seems to be incompatible with the latest one used in the toolbox.)

**Comment by rbodo** [comment]
If you are going to use a new keras model (other than the example provided with the toolbox), see #25

**Comment by piewchee** [comment]
> 
> 
> Did you try with a longer simulation time (e.g. 100 time steps)? This should give you good accuracy on the LeNet example with Nest simulator. (Even though the ANN accuracy is still low - which is due to the model being trained with an old Keras version that seems to be incompatible with the latest one used in the toolbox.)

Hi Bodo,

After I changed the simulation time to 100.
INI attained 100% for both ANN and SNN.
nest attained 100% ANN 90% SNN.

Rgds
Del

**Comment by rbodo** [comment]
ah, with which keras version?

**Comment by piewchee** [comment]
> 
> 
> ah, with which keras version?

I am using 2.1.6 followed your recommendation.

By the way, I am modifying the scripts to try out interfacing with spinnaker board.
Do I change the scripts (target simulator, config.defaults etc.) in the snn_toolbox in my working directory
or the snn_toolbox in python2.7/site_packages directory?
It seems the simulator is reading the scripts from  python2.7/site_packages.

Rgds
Del

**Comment by rbodo** [comment]
> I am using 2.1.6 followed your recommendation.

OK, that explains the high ANN accuracy.
 
> By the way, I am modifying the scripts to try out interfacing with spinnaker board.
> Do I change the scripts (target simulator, config.defaults etc.) in the snn_toolbox in my working directory
> or the snn_toolbox in python2.7/site_packages directory?
> It seems the simulator is reading the scripts from python2.7/site_packages.

I guess that depends on how you installed the toolbox...
I would clone or fork the repo and install with `python setup.py develop` or `pip install -e .`. Then you can just edit the cloned repo files and the changes will take effect also in the site_packages directory.

**Comment by piewchee** [comment]
> 
> 
> > I am using 2.1.6 followed your recommendation.
> 
> OK, that explains the high ANN accuracy.
> 
> > By the way, I am modifying the scripts to try out interfacing with spinnaker board.
> > Do I change the scripts (target simulator, config.defaults etc.) in the snn_toolbox in my working directory
> > or the snn_toolbox in python2.7/site_packages directory?
> > It seems the simulator is reading the scripts from python2.7/site_packages.
> 
> I guess that depends on how you installed the toolbox...
> I would clone or fork the repo and install with `python setup.py develop` or `pip install -e .`. Then you can just edit the cloned repo files and the changes will take effect also in the site_packages directory.

Alrite Thanks!!
Will try and see.

Rgds
del

**Comment by piewchee** [comment]
I did a minor modification to the scripts and tried to run the snntoolbox using simulator=spiNNaker.
It returns the error below: not sure what is the problem.
Is it trying to read the spiNNaker projection label to form the pathname?
rgds
Del

Detected layer with biases: 00Conv2D_6x24x24
Detected layer with biases: 02Conv2D_16x8x8
Detected layer with biases: 04Conv2D_120x4x4
Detected layer with biases: 06Dense_84
Detected layer with biases: 07Dense_10
Number of operations of ANN: 2346734
Number of neurons: 7614
Number of synapses: 1397800

Saving model to /media/mikat3/92d61e88-74ff-49dd-8e8b-5f1ea9ec3eb7/home/mikat2/snn_toolbox/examples/models/lenet5/keras...
Saving assembly...
Saving connections...
Traceback (most recent call last):
  File "/usr/local/bin/snntoolbox", line 9, in <module>
    load_entry_point('snntoolbox', 'console_scripts', 'snntoolbox')()
  File "/media/mikat3/92d61e88-74ff-49dd-8e8b-5f1ea9ec3eb7/home/mikat2/snn_toolbox/snntoolbox/bin/run.py", line 50, in main
    test_full(config)
  File "/media/mikat3/92d61e88-74ff-49dd-8e8b-5f1ea9ec3eb7/home/mikat2/snn_toolbox/snntoolbox/bin/utils.py", line 116, in test_full
    config.get('paths', 'filename_snn'))
  File "/media/mikat3/92d61e88-74ff-49dd-8e8b-5f1ea9ec3eb7/home/mikat2/snn_toolbox/snntoolbox/simulation/target_simulators/pyNN_target_sim.py", line 174, in save
    self.save_connections(path)
  File "/media/mikat3/92d61e88-74ff-49dd-8e8b-5f1ea9ec3eb7/home/mikat2/snn_toolbox/snntoolbox/simulation/target_simulators/pyNN_target_sim.py", line 356, in save_connections
    filepath = os.path.join(path, projection.label.partition('â†’')[-1])
AttributeError: 'Projection' object has no attribute 'label'

**Comment by rbodo** [comment]
Yes, in, pyNN, each projection has a label, typically something like `source_population_name->target_population_name`. In line 356, I extract the part of this label behind the arrow and use it as filename of the stored connections. It seems your new SpiNNaker projection object does not have such a label. That's an easy fix though, either find the equivalent attribute of your object, or invent some label for each connection to store.

**Comment by piewchee** [comment]
hi Bodo,
Wouldn't it be more appropriate if spiNNaker_target_sim follow the same naming convention? cause the population connection will load the weight, delay info from the according to the name.
If I change the label in different format, will that mess thing up?
Rgds
Del

**Comment by rbodo** [comment]
You are right. I just don't know anything about your SpiNNaker projection object. Is that something you created yourself? The easiest would be to just give it the `label` attribute and set it with the `target_population_name` as in the pyNN simulators.

**Comment by piewchee** [comment]
I didn't change much thing in your scripts. As pyNN_target_sim.py is very similar to pyNN.spinnaker; hence I just copy and rename it to spinnaker_target_sim.py, modified the utils.py and config.defaults's restrictions.
I find it strange and still trying to figure out why your 
"filepath = os.path.join(path, projection.label.partition('->')[-1]) works
ie. it can find the projection label.
I checked your connections function, the projection did not have any label declared (pyNN will auto-generate label if not declared). 
I mean why does it not complain that projection has no label attribute and is able to rename it to the 'target population name' using ('->')[-1].  
your assembly(population) function do have label named after the layer.
Sorry I am not a software guy, more of IC designer, so not very familiar with scripting

Rgds
Del

**Comment by rbodo** [comment]
> I checked your connections function, the projection did not have any label declared (pyNN will auto-generate label if not declared).
Exactly, this is what's happening: I never set the label explicitly, so pyNN autogenerates the label to `source_population_name->target_population_name`. It seems the pyNN.spinnaker module does not autogenerate a label, hence the error. 

> I mean why does it not complain that projection has no label attribute and is able to rename it to the 'target population name' using ('->')[-1].
It does not complain because the label is autogenerated. I do not rename the label. I _read_ the label and extract the second part (removing the arrow because it's not a standard character) and use this part to name my file where I save the connections.

You could create the filename yourself, with something like:

```
for i, projection in enumerate(self.connections):
    filename = projection.label.partition('→')[-1] if hasattr(projection, 'label') else 'layer_' + str(i)
    filepath = os.path.join(path, filename)
    ...
```

**Comment by piewchee** [comment]
Hi Bodo,
The cell(neuron) parameters that you set in config.defaults : eg v_thresh=1mv, v_reset=0 etc
won't change throughout the inference.
May I know how do you determine what values to set?

Rgds
Del

**Comment by rbodo** [comment]
`v_thresh` is set to 1 mV so that the normalized network can fire at most one spike per 1 ms time step. You could set the threshold differently and adapt the normalization accordingly.
`v_reset` is zero for simplicity and because it corresponds to the way the original ANN runs. If you change the reset value, you are essentially introducing a global bias towards higher or lower firing rates.
Not all of the other parameters may actually be used by the neuron model. Those that are were determined experimentally such that the spike rates fit most closely the activations of the original ANN. We just want a simple integrate-and-fire neuron with instantaneous integration and without any leak or delay.

**Comment by piewchee** [comment]
> 
> 
> `v_thresh` is set to 1 mV so that the normalized network can fire at most one spike per 1 ms time step. You could set the threshold differently and adapt the normalization accordingly.
> `v_reset` is zero for simplicity and because it corresponds to the way the original ANN runs. If you change the reset value, you are essentially introducing a global bias towards higher or lower firing rates.
> Not all of the other parameters may actually be used by the neuron model. Those that are were determined experimentally such that the spike rates fit most closely the activations of the original ANN. We just want a simple integrate-and-fire neuron with instantaneous integration and without any leak or delay.

Hi Bodo,
After sometimes, we finnaly managed to run the snntoolbox using Spinnaker as simulator.
However, I am unable to get the accuracy correctly.
The neuron parameters are exactly the same as using "nest" simulator and Vthr = 1mV.
I run the Lenet5/keras model (mnist) using "nest" first to get all the layers synaptic weight parameters.
After that I run using "spinnaker" with the synaptic weights previous saved using "nest".

However, I noticed that the number of spikes for each layers are not as similar to the "nest" output.
some of the test sample output has no spike output at all.
eg. the truth_d and guessed_d  for 10 samples as follow:
truth_d =[7,2,1,0,4,1,4,9,5, 9]
guessed_d = [-1, -1, 0, 0, -1, 0, 0, 0, 0 , -1] 

Not sure if you are familiar with Spinnaker, is there any significant difference between nest neuron and spinnaker?

Rgds
Del

**Comment by rbodo** [comment]
Hi Del,

Congrats on getting the network to run on Spinnaker.

If the problem is low output rates, did you try the following?

- Make sure there is some output at all (this seems to be the case here judging from some of the samples being assigned class 0 instead of -1 which corresponds to no activity)
- Use longer simulation times and/or finer time resolution to give the neurons time to integrate enough input for producing a spike.
- Alternatively, lower the threshold. Also, play with other neuron parameters, e.g. make sure the leak is turned off.
- Use artificial input with extreme parameter settings to enforce some output, for instance a completely white image (all pixels 255), very low threshold, very long simulation time.
- Inspect the membrane potentials of hidden layer neurons to see that they behave as expected.
- Try with fewer layers first and add them one by one to pinpoint the problem.

Good luck!

**Comment by piewchee** [comment]
Hi Bodo,
> 

>     * Use longer simulation times and/or finer time resolution to give the neurons time to integrate enough input for producing a spike.

I have already tried increasing the simulation times to 200ms or more (or lower the Vthres). However the
spinnaker will face back pressure congestion as there are too many spikes occur at the same times, hence many packets are being dropped.
> 
>     * Alternatively, lower the threshold. Also, play with other neuron parameters, e.g. make sure the leak is turned off.
I tried this too,  i lowered the threshold to 0.85, i looked at the total_spikes_activity.png plot, I noticed the number of spikes is higher than when using "nest" simulator, if I increase it to 0.9 it will be much lower
than what "nest" produced. 
Hence the layers are able to generate spikes, just that the guessed output is wrong. 
In "config_default" file, "leak=False", so leak is already turn off?

Will try the following methods:
>     * Use artificial input with extreme parameter settings to enforce some output, for instance a completely white image (all pixels 255), very low threshold, very long simulation time.
> 
>     * Inspect the membrane potentials of hidden layer neurons to see that they behave as expected.
> 
>     * Try with fewer layers first and add them one by one to pinpoint the problem.
> 
> 

Rgds
Del

**Comment by piewchee** [comment]
> 
> 
> > In "config_default" file, "leak=False", so leak is already turn off?
> 
> I meant turning off the leak on Spinnaker side. That would mean choosing a very long time constant for the membrane potential decay. The corresponding parameter in the nest simulator is `tau_m`, which I set to 1000.

Hi Bodo,

Spinnaker is using the same neuron parameters, hence tau_m=1000 is used.

I ran a simple 3 neurons in a loop connection(synfire chain) with SpikeSourceArray used as stimulus
connected to neuron(0). 
Attached figure 3 and figure 4 are the results from Nest and Spinnaker outputs. The neuron behaviours
look similar, except that NEST generates the spiketrain one time step after the neuron voltage potential
hit Vthreshold, whereas Spinnaker generate the spiketrain at the same timespot when Vmem hit Vthre

I replaced the stimulus to SpikeSourcePoisson for the synfire chain with poisson rate=50, as shown in Fig nest50 and spin50 (for nest and spinnaker simulators). I understand that it's a random generator. however the nest simulator generates much lesser spikes than Spinnaker consistently for a number of simulation runs.

If I am not wrong, the poisson rate for the input layer is scaled to 100  max?
neuron.rate = rates[neuron_idx] / self.rescale_fac * 1000
Could that be the reason I am unable to get the test accuracy for spinnaker compared with NEST?
![figure_3](https://user-images.githubusercontent.com/31268312/51100788-a77d5e80-1812-11e9-9ea6-d2dfd11fc210.png)
![figure_4](https://user-images.githubusercontent.com/31268312/51100806-b5cb7a80-1812-11e9-8220-340557f9ded9.png)
![nest50](https://user-images.githubusercontent.com/31268312/51100814-bd8b1f00-1812-11e9-90ce-5e8691bb5d66.png)
![spin50](https://user-images.githubusercontent.com/31268312/51100818-c3810000-1812-11e9-8d02-12b132c4e7ae.png)


Rgds
Del

**Comment by piewchee** [comment]
Another observation when I use SpikeSourcePoisson as stimulus.
For NEST simulator, the synfire chain is consistently output same voltage potentials and spikes for any number of simulations i run (ie the output plot nest50.png
is exactly the same no matter how many times I run as long as the poisson rate is same).
For Spinnaker, the output plot is different for each simulation run.
In conclusion: Spinnaker PoissonSource generates more spikes and each run produces different number of spikes.

Rgds
Del

**Comment by rbodo** [comment]
> For NEST simulator, the synfire chain is consistently output same voltage potentials and spikes for any number of simulations i run (ie the output plot nest50.png
> is exactly the same no matter how many times I run as long as the poisson rate is same).
> For Spinnaker, the output plot is different for each simulation run.

I guess the reason for this is that we fix the random seed for reproducibility when using NEST as simulator backend.

**Comment by rbodo** [comment]
> If I am not wrong, the poisson rate for the input layer is scaled to 100 max?
> neuron.rate = rates[neuron_idx] / self.rescale_fac * 1000

The Poisson rate can be set with the SNN toolbox config file:

```
[input]
input_rate = 1000
```

If you set the ``input_rate`` parameter to 1000, then an input image pixel with maximum value (1 or 255) will fire on average at every time step. Lowering ``input_rate`` lowers the Poisson spike rate. So you can tune this parameter to try and get the same behavior between NEST and Spinnaker.

**Comment by piewchee** [comment]
Hi Bodo,
I think is the NEST simulator that fixed the NRG, as I wasn’t using the snntoolbox to run the synfire chain. I ran it using the standalone nest simulator.

Rgds
Del

**Comment by piewchee** [comment]
H Bodo,

My bad, the default scaling factor is 1000, I changed it to 100 cause I thought the high fire rate causes the Spinnaker to have
Congestion.

If I use scaling factor of 1000, both NEST and Spinnaker have similar firing pattern.

However, no matter I tune the scaling factor to lower / higher than 1000 by 10 folds or more, the Spinnaker doesn’t give me better results.

Best Regards
Del

**Comment by piewchee** [comment]
Hi Bodo,

Is it okay if I send the script to Spinnaker team to figure out what is wrong between Nest and Spinnaker simulator?

Best Regards
Del

**Comment by piewchee** [comment]
> 
> 
> Yes, absolutely, would be interested to learn the reason for this discrepancy.

Hi Bodo,
neuron.rate = rates[neuron_idx] / self.rescale_fac * 1000
If I am not wrong, the dataset (neuron_idx) parameter is normalise to 0-1 (ie. 255=1?)
Hence for me to simulate all pixel = 255 or 0;
I set neuron.rate=1000 or neuron.rate=0.
For NEST simulator, I noticed that there is no output spike for the last layer eventhough there are spiketrains in the inner layers.

Whereas for Spinnaker, there are spike outputs for the last layer when I set neuron.rate=1000,
for neuron.rate = 0, the system automatically set the neuron.rate =1.

Rgds
Del

**Comment by rbodo** [comment]
Yes, that's because it does the whole conversion from scratch, using the Keras .h5 model. In some cases (e.g. INI simulator) it is possible to just start with loading the converted SNN, but if I remember correctly that was not feasible for pyNN simulators, so the toolbox runs the whole pipeline from the beginning.

**Comment by piewchee** [comment]
Hi Bodo,

I have the feeling that I have got the spinnaker working correctly.
However I need further tests.

As mentioned to you before, the input test pattern that used to generate poisson input is different between NEST and Spinnaker.
NEST has fixed the seed number for the poisson rate generator, whereas Spinnaker using RNG.

In this case, I have used NEST simulator to generate the input poisson pattern and feed into spinnaker. The down side is
that I can only generate the poisson rate based on 1 test digit.

As far as I know, the test set always generate same number sequence. The first test digit is 7.
I have run 1 digit test (ie. 7, poisson rate from NEST) using Spinnalker and ran a few times, it always guess corrently. The Pearson plot
Also looks promising.

Now I would like know if there is anyway to change the first test digit to other numbers?

I have tried using poisson rate genearated by Spinnaker, it still fails. I will further probe into this later.

Best Regards
Del

**Comment by rbodo** [comment]
Congratulations on your progress!

Not sure I understand your question / how you generate the poisson rates from NEST. If you stored the original MNIST digits in an npz file, you should be able to simply save a version of this file with the order of the digits permuted. Or just permute the data in python after loading it from disk and before feeding it to your spike generator. Or maybe I'm missing something?

**Comment by piewchee** [comment]
Hi Bodo,
What I mean is that the input layer where you are using SpikeSourcePossion from NEST, the seed number is fixed; however for Spinnaker the
Seed number is generated using RNG. Hence every simulation run, NEST simulator always produce identical spiketrain pattern from the input layer for
Digit 7, whereas Spinnaker has different spiketrain patterns for the same digit for every run.

So what I did is I saved the spiketrains pattern for the input layer from Nest Simulation, and replaced the SpikeSourcePoisson with SpikeSourceArray
And use this spiketrain in Spinnaker.

What I can do in order to get different digit is to simulate using different number of num_to_test, as the script only save the last num_to_test spiketrains.

Best Regards
Del

**Comment by rbodo** [comment]
Ah, thanks for clarifying.

Yes, you could do as you suggested. Another option is to use the parameter 

```
[simulation]

sample_idxs_to_test = [0, 5, 2123]
```

in the config file.

This allows you to specify which sample of the testset you want to run (0, 5, 2123 in the example). I hope this still works for NEST, haven't used it in a long time.

But anyways, the toolbox saves the spiketrains of hidden layers for every sample, if you specify in the config file

```
[output]
log_vars = {'spiketrains_n_b_l_t', 'input_b_l_t', ...}
```

The data is stored in the directory 

```
[paths]
log_dir_of_current_run =
```
or a subfolder of the working directory if the path above is not specified.

If I remember correctly, the `spiketrains_n_b_l_t` variable does not include the Poisson input spiketrains, but you should be able to easily add that to the existing logging behavior.

**Comment by piewchee** [comment]
Hi Bodo,
Yes I have tried the sample_idxs_to_test = [1,4,7,9]
I am not sure what should I set the num_to_test in config. Anyway I set it =1
When I run the nest simulation.
It gives me this error msg.
x_test = np.array([x_test[i] for I in si])
Index Error: index 4 is out of bounds for axis 0 with size 4.

Anyway for Spinnaker that uses SpikeSorucePoisson,
I suspect that Spinnaker simulator keep using the neuron.rate  from the first test digit ie [7].
I am not sure why it does not read in the new test digit when it is starting a new simulation.
I print out neuron.rate , and it does show a new neuron.rate is printed out for each new simulation with respect to the a new test digit.

Best Regards
Del

**Comment by piewchee** [comment]
Hi Bodo,

The Spinnaker toolbox finally work. I run 20 test samples and 3 were wrong, hence acc=85%.
I changed the V_thres =0.75

Apparently Spinnaker SpikeSourcePoisson() has some issue;
in NEST, neuron.rate will set the rate for SpikeSourcePoisson, and the rate is updated whenever there is a new test digit.

However , Spinnaker keep repeating the same first test digit neuron.rate.

Spinnaker guy advised me to use layer.set(rate=..) instead; so I changed the self._poisson_input:
rates= kwargs[str('x_b_l')].flatten()
spiketrains = []
spiketrains = list(rates*500)  # I reduced the scaling for spinnaker)
self.layers[0].set(rate=spiketrains)
 
Rgds
Del

**Comment by piewchee** [comment]
Hi Bodo,
I am thinking of trying out other models, eg. cifar10, on Spinnaker
However when I look into the config file in binarynet , the simulator option is not stated in there, ie it is
using config.defaults simulator which is INI.
May I know if I can include "simulator=nest, etc" in the config file so that SNN can be simulated
using nest, spynnaker etc.

Rgds
Del

**Comment by rbodo** [comment]
BinaryNet has binary weights and binary activations. Binary weights would not be a problem because the toolbox ensures the weights are binarized during parsing of the input model, if you specify it in the config file. There is nothing in the nest simulator that ensures binary activations, so I don't know whether it will come out all right. You can try of course. But you should be aware that the BinaryNet is rather large, so it may take a while to run.

An alternative is to use a standard CNN architecture for CIFAR10; keras has a few examples.

https://github.com/keras-team/keras/blob/master/examples/cifar10_cnn.py

You may want to train it with Averagepooling and without biases because they are not implemented in Nest.

**Comment by piewchee** [comment]
Thanks!

**Comment by piewchee** [comment]
H Bodo,

I have re-training cifar10_cnn.py disabling the bias in CNNlayer (use_bias=False) and averagepooling.

I ran the nest simulator, the toolbox killed the process while saving layer 01conv. not sure what happened. I set normalized = True, and binarized_weights= False.
(i changed the dataset path to "datasets/binaryconnect")

Anyway since saving the layer connections has no effect on the simulation, i disabled it.
I encountered another problem after that, the poisson generator encountered negative rate.
Am I using the right dataset? binaryconnect dataset has normalised data, but why is there negative
values?

Best Regards
Del

**Comment by rbodo** [comment]
If I remember correctly, the dataset was preprocessed by the authors of BinaryConnect (whitened, mean-subtracted etc), which would explain the negative values. 

In the training script, you can save your own version of the cifar10 dataset without mean-subtraction and use it in the simulator.

**Comment by piewchee** [comment]
Hi Bodo,

Sure. I need to tidy up the script abit. There are many debug lines in the script need to be cleaned up.
Let me try to settle the cifar10 problems first.

Best Regards
Piew Yoong

**Comment by piewchee** [comment]
Hi Bodo,

A quick question.
In the config file. If I set  under [tools]- normalise=True.
Then the dataset will require to have x_norm.npz and x_test.npz.
If normalise=False, then only x_test.npz is required.

If I am not wrong the normalisation is for weight normalisation.

What setting do I need to change if I run Keras model training to obtain  the .h5 file.

Rgds
Del

**Comment by rbodo** [comment]
> In the config file. If I set under [tools]- normalise=True. Then the dataset will require to have x_norm.npz and x_test.npz. If normalise=False, then only x_test.npz is required. 

Correct.

> If I am not wrong the normalisation is for weight normalisation. 

Correct.

> What setting do I need to change if I run Keras model training to obtain the .h5 file.

I would not use the toolbox (i.e. the config file) to train a Keras model. Instead you can start with the script 

`snn_toolbox/scripts/ann_architectures/cifar10/cnn_BN.py` 

or use some of the others in that folder.

**Comment by rbodo** [comment]
> Sure. I need to tidy up the script abit. There are many debug lines in the script need to be cleaned up. Let me try to settle the cifar10 problems first.

That's great! (By the way, the script does not need to be perfect since it is modular and should not affect other parts of the toolbox, so we can keep working on it without fear of breaking things.)

**Comment by piewchee** [comment]
Hi Bodo,

Yes I am using the ann_architecture models for training.
What I don’t understand is what shall I change in cnn_BN.py in order to normalise weight?
How does that affect the dataset? (ie we need to generate x_norm.npz)

Best Regards
Del

**Comment by piewchee** [comment]
Hi Bodo,

By the way, as I am new to this.
I wish to rescale the dataset pixel density between 0 – 1.

As I mentioned in previous post, I am not sure how to generate x_norm.npz data.
May I know how to  save the x_test.npz and y_test.npz files as well as x_norm.npz?

Not sure if my script is correct for saving the x_test.npz.

(x_train, x_train), (x_test, y_test) = cifar10.load_data()
----

test_datagen = ImageDataGenerator(rescaled =1./255, featurewise_centre=False, featurewise_std_normalization=False,
zca_whitening = False)
xnp =[]
num_xnp=0
for x_batch, y_batch in test_datagen.flow(x_test, y_test):
     xnp.append(x_batch)
     num_xnp +=1
     if num_xnp == x_test.shape[0]:
        break
xnp = xnp.concatenate(xnp)
np.savez_compressed(save_dir + '/x_test', xnp)

Rgds
Del

**Comment by rbodo** [comment]
`x_norm.npz` is just a subset or a copy of the dataset. If the dataset is as small as cifar10, you can simply copy `x_test.npz`.

To rescale the dataset, just divide by 255; there is no need to use the `ImageDataGenerator` except you want more sophisticated preprocessing or your dataset does not fit into memory.

There is an example of how to save the dataset for the toolbox here:

`/snn_toolbox/scripts/dataset_io/load/cifar10.py`

(You can turn off global contrast normalization though.)

**Comment by delpie** [comment]
Hi Bodo,

Thanks! I used the cifar10.py to save the x_test.npz etc
It has the same error that I encountered when I using my own script to save the x_test, y_test.npz
the error is as follow:
File "/media/mikat3/92d61e88-74ff-49dd-8e8b-5f1ea9ec3eb7/home/mikat2/snn_toolbox/snntoolbox/simulation/utils.py", line 1069, in reshape_flattened_spiketrains
    spiketrains_flat[k, int(t / self._dt)] = t
IndexError: index 784 is out of bounds for axis 0 with size 784

Rgds
Del

**Comment by rbodo** [comment]
Hi Del,

Please check your input dimensions in the model definition. The cifar10 images are of shape `(32, 32, 3)`. MNIST has `np.prod((28, 28, 1)) = 784`, so my guess is that you still have the MNIST input dimension in your network.

**Comment by piewchee** [comment]
Hi Bodo,

Thanks! My bad! I forgot that I have hardcoded the input dimension when I was debugging MNIST test.

Best Regards
Del

**Comment by piewchee** [comment]
Hi Bodo,

I tried to access your github, but was denied.
Permission to neuromorphicprocessorproject/snn_toolbox…. Denied to piewchee

Rgds
Del

**Comment by rbodo** [comment]
What exactly do you mean with "access"? If you want to upload something (issuing a pull request), you have to do this from a fork of the repository (because snntoolbox is part of an organization account, so only members of the organization can make pull requests from a clone of the repo.) See also here:

https://help.github.com/en/articles/creating-a-pull-request-from-a-fork

So I guess you'd fork the toolbox, merge your changes into the fork, and then submit a pull request.

Or did you mean something else?

**Comment by piewchee** [comment]
Hi Bodo,
I did

1.       git add .

2.       git commit –m spynnaker8_target_sim.py

3.       git push snn_toolbox/snntoolbox/simulation/target_simulators spynnaker8_target_sim.py

Rgds
Del

**Comment by rbodo** [comment]
I see. Yes, that won't work. Please use a pull request as I described above.

(If it's too much trouble, you can also send me the file and I'll upload it since it's only the one file.)

**Comment by piewchee** [comment]
Hi Bodo,

Thanks! I will send you the file then, cause previously you asked me to push it to the repository, so I thought
It was allowed.

Rgds
Del

**Comment by rbodo** [comment]
Ah, sorry for the misunderstanding. Do you have my email? rbodo@ini.uzh.ch

**Comment by rbodo** [comment]
Del asked me to share the following modifications he had to do to run a model converted with the snntoolbox on SpiNNaker.

1. In `sPyNNaker/spynnaker/pyNN/models/pynn_population_common.py` inside `def get` function, add the following statement in line 203:

```
if isinstance(parameter_names, unicode):
    parameter_names=str(parameter_names)
```

2. In `sPyNNaker/spynnaker/pyNN/models/neuron/builds/if_cond_exp_base.py`, add

```
def describe(self):
    return “IF_cond_exp”
```

3. In `sPyNNaker/spynnaker/pyNN/models/spike_source/spike_source_poisson.py`, add 

```
def describe(self):
    return “SpikeSourcePoisson”
```

4. In `SpiNNUtils/spin_utilities/ranged/abstract_dict.py`, inside `def __contains__` function, add in line 221:

```
if isinstance(key, unicode):
    key = str(key)
```

5. In `sPyNNaker8/spynnaker8/models/projection.py`, give the `data` argument the default value `data=None` in

```
def _save_callback(self, save_file, format, metadata, data):
````

**Comment by Aashish3970** [comment]
Hello,
I tried using the MNIST keras spinnaker example https://github.com/NeuromorphicProcessorProject/snn_toolbox/blob/master/examples/mnist_keras_spiNNaker.py. When I deploy this on spinnaker, the moving accuracy of ANN is 100% the moving accuracy of SNN is 0%. After that, I changed the simulation duration to 150 and v_thresh to 0.75. I am still getting the same accuracy. Plus, I get Average spike rate :0 spikes per simulation time step. What am I doing wrong?
![output](https://user-images.githubusercontent.com/54250712/94970372-2a07ef00-04ca-11eb-89cd-fce63a4f7500.png)

Below is my config file
![image](https://user-images.githubusercontent.com/54250712/94970563-8b2fc280-04ca-11eb-81a6-b546d1887c2e.png)
